trio/_core/_run.py

```diff
@@ -903,13 +903,18 @@ def __enter__(self) -> None:
             "use 'async with open_nursery(...)', not 'with open_nursery(...)'"
         )
 
-    def __exit__(self) -> None:  # pragma: no cover
+    def __exit__(
+        self,  # pragma: no cover
+        exc_type: type[BaseException] | None,
+        exc_value: BaseException | None,
+        traceback: TracebackType | None,
+    ) -> bool:
```

Comment:
This along with `__enter__` should probably be hidden behind `if not TYPE_CHECKING`, so type checkers will correctly just give an error if used in a regular sync `with` statement. Potentially also the return types could be changed to `typing.NoReturn`, for runtime checkers perhaps.`__exit__()` will need to be swapped to `raise AssertionError(...)`, which is probably better anyway (since `-O` would cause it to just return `None`).